### PR TITLE
[SPARK-39877][PYTHON] Add unpivot to PySpark DataFrame API

### PR DESCRIPTION
**Pull request by EnricoMi**
### What changes were proposed in this pull request?
This adds `unpivot` and its alias `melt` to the PySpark API. It calls into Scala `Dataset.unpivot` (#36150). Small difference to Scala method signature is that PySpark method has default values. This is similar to `melt` in Spark Pandas API.

### Why are the changes needed?
To support `unpivot` in Python.

### Does this PR introduce _any_ user-facing change?
Yes, adds `DataFrame.unpivot` and `DataFrame.melt` to PySpark API.

### How was this patch tested?
Added test to `test_dataframe.py`.

### CODE REVIEW

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -2188,6 +2188,150 @@ def cube(self, *cols: "ColumnOrName") -> "GroupedData":  # type: ignore[misc]
...
+            valueColumnName = "value"
+
+        return DataFrame(
+            self._jdf._unpivot(
+                self._jcols(*id_list), self._jcols(*value_list), variableColumnName, valueColumnName
+            ),
+            self.sparkSession,
+        )
+
+    def melt(
```
Comment (zhengruifeng) [inline python/pyspark/sql/dataframe.py:2294]: there was also a `melt` function in `python/pyspark/pandas/frame.py`, is it possbile to apply the new `unpivot` implementation there?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -2127,6 +2127,15 @@ class Dataset[T] private[sql](
       valueColumnName: String): DataFrame =
     unpivot(ids, Array.empty, variableColumnName, valueColumnName)
 
+  /**
+   * Called from Python as Seq[Column] are easier to create via py4j than Array[Column].
+   */
+  private[sql] def _unpivot(ids: Seq[Column],
```
Comment (zhengruifeng) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2133]: I guess we can move it to `src/main/scala/org/apache/spark/sql/api/python/PythonSQLUtils.scala`

File: python/pyspark/sql/dataframe.py (hunk 1 above)
Comment (EnricoMi) [reply python/pyspark/sql/dataframe.py:2294]: That is planned, but I'd prefer to do that in a separate PR: SPARK-39878

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (EnricoMi) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2133]: I'd prefer to get `PythonSQLUtils.toArray` return array of `Column`s but it returns array of `Object`. Looks like generic types do not work as expected.

Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2133]: maybe we can directly change the types of `ids` and `values` from `Array[Column]` to `Seq[Column]`?

Comment (EnricoMi) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2133]: The `Array[Column]` was chosen over `Seq[Column]` to make `Dataset.unpivot` Java-friendly.

File: python/pyspark/sql/dataframe.py [hunk 3]
```diff
@@ -2188,6 +2188,142 @@ def cube(self, *cols: "ColumnOrName") -> "GroupedData":  # type: ignore[misc]
...
+            if cols is None:
+                l = []
+            elif isinstance(cols, tuple):
+                l = list(cols)
+            elif not isinstance(cols, list):
+                l = [cols]
+            return self._jcols(*l)
+
+        if variableColumnName is None:
+            variableColumnName = "variable"
```
Comment (zhengruifeng) [inline python/pyspark/sql/dataframe.py:2274]: default values `variableColumnName = "variable"` and `valueColumnName = "value"` are only used in pyspark?

Scala side doesn't have similar logic?

Comment (itholic) [reply python/pyspark/sql/dataframe.py:2274]: > Scala side doesn't have similar logic?

Seems like Scala side doesn't have similar logic according to the [implementation](https://github.com/apache/spark/pull/36150/files) ?

Comment (EnricoMi) [reply python/pyspark/sql/dataframe.py:2274]: The rational in Scala was that other SQL systems require variable and value column names (there is no default), so we decided Scala API does not have default values either. Given that PySpark Pandas API has default values I thought PySpark Dataframe API could have one too.

If you prefer PySpark Dataframe API to match Scala Dataset API rather than PySpark Pandas API, I am fine with that.

File: python/pyspark/sql/dataframe.py [hunk 4]
```diff
@@ -2188,6 +2188,142 @@ def cube(self, *cols: "ColumnOrName") -> "GroupedData":  # type: ignore[misc]
 
         return GroupedData(jgd, self)
 
+    def unpivot(
+        self,
+        ids: Optional[Union["ColumnOrName", List["ColumnOrName"], Tuple["ColumnOrName", ...]]] = None,
+        values: Optional[Union["ColumnOrName", List["ColumnOrName"], Tuple["ColumnOrName", ...]]] = None,
```
Comment (EnricoMi) [inline python/pyspark/sql/dataframe.py:2194]: What are your thoughts on allowing for tuple of columns as an alternative to list of columns (equivalent semantics)? `Dataframe._jcols` does not support tuples, only list and single string so this does not seem to be standard behaviour.

Comment (zhengruifeng) [reply python/pyspark/sql/dataframe.py:2194]: I'm netural on it.

some existing methods also use tuple of column names:

```
    def freqItems(
        self, cols: Union[List[str], Tuple[str]], support: Optional[float] = None
    ) -> "DataFrame":
```

File: python/pyspark/sql/dataframe.py (hunk 3 above)
Comment (zhengruifeng) [reply python/pyspark/sql/dataframe.py:2274]: i think we'd better keep the behaviors the same.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2133]: got it.

since this method is still needed and dedicated only for python, what about move it to `PythonSQLUtils`?

something like

```
def unpivot(dataframe: DataFrame, ids: Seq[Column] ...) = {
    dataframe.unpivot(ids.toArray, values.toArray, variableColumnName, valueColumnName)
}

```

**Comment by zhengruifeng** [comment]
btw, you may also need to run `dev/reformat-python`

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (EnricoMi) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2133]: Sure, just hesitating because that looks like a very specific method for such a generic utils class. The other methods are more generic. And there are other private Python methods in `Dataset` (`mapInPandas`) and `RelationalGroupedDataset` (`flatMapGroupsInPandas`).

File: python/pyspark/sql/dataframe.py (hunk 3 above)
Comment (EnricoMi) [reply python/pyspark/sql/dataframe.py:2325]: removed

**Comment by EnricoMi** [comment]
> btw, you may also need to run `dev/reformat-python`

Why do I have to reformat `python/pyspark/context.py`? That seems unrelated.

File: python/pyspark/context.py [hunk 5]
```diff
@@ -309,10 +309,7 @@ def _do_init(
         if sys.version_info[:2] < (3, 8):
             with warnings.catch_warnings():
                 warnings.simplefilter("once")
-                warnings.warn(
-                    "Python 3.7 support is deprecated in Spark 3.4.",
-                    FutureWarning
-                )
+                warnings.warn("Python 3.7 support is deprecated in Spark 3.4.", FutureWarning)
```
Comment (zhengruifeng) [inline python/pyspark/context.py:312]: I encounter the same issue in other PR, maybe due to the `black` version. Let us revert this change

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2114]: can you rename the method? we do not use `_xxx` like function names in scala

Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2114]: > And there are other private Python methods in Dataset (mapInPandas) and RelationalGroupedDataset (flatMapGroupsInPandas).

they are also used in test

File: python/pyspark/context.py (hunk 5 above)
Comment (EnricoMi) [reply python/pyspark/context.py]: done

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (EnricoMi) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Renamed `private[sql] def _unpivot` to `private[sql] def unpivot`.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 6]
```diff
@@ -2107,6 +2107,16 @@ class Dataset[T] private[sql](
...
     unpivot(ids, Array.empty, variableColumnName, valueColumnName)
 
+  /**
+   * Called from Python as Seq[Column] are easier to create via py4j than Array[Column].
+   * We use Array[Column] for unpivot rather than Seq[Column] as those are Java-friendly.
+   */
+  private[sql] def unpivotWithSeq(ids: Seq[Column],
+                                  values: Seq[Column],
+                                  variableColumnName: String,
+                                  valueColumnName: String): DataFrame =
```
Comment (zhengruifeng) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2117]:
```suggestion
  private[sql] def unpivotWithSeq(
      ids: Seq[Column],
      values: Seq[Column],
      variableColumnName: String,
      valueColumnName: String): DataFrame =
```

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
Merged to master, thanks @EnricoMi for working on this!

File: python/pyspark/sql/dataframe.py [hunk 7]
```diff
@@ -2238,6 +2238,140 @@ def cube(self, *cols: "ColumnOrName") -> "GroupedData":  # type: ignore[misc]
 
         return GroupedData(jgd, self)
 
+    def unpivot(
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py:2241]: Let'd document it under https://github.com/apache/spark/blob/master/python/docs/source/reference/pyspark.sql/dataframe.rst

**Comment by HyukjinKwon** [comment]
LGTM2

File: python/pyspark/sql/dataframe.py (hunk 7 above)
Comment (EnricoMi) [reply python/pyspark/sql/dataframe.py:2241]: https://github.com/apache/spark/pull/37354
